core/src/main/java/org/apache/struts2/interceptor/exec/StrutsBackgroundProcess.java

```diff
@@ -61,11 +66,17 @@ public BackgroundProcess prepare() {
                 try {
                     beforeInvocation();
                     result = invocation.invokeActionOnly();
-                    afterInvocation();
                 } catch (Exception e) {
+                    LOG.warn("Exception during invokeActionOnly() execution", e);
                     exception = e;
                 } finally {
-                  done = true;
+                    try {
+                        afterInvocation();
+                    } catch (Exception ex) {
+                        exception = ex;
```

Comment:
Hmm what if `beforeInvocation()` already threw an exception, do we want to overwrite that?

Reply:
Yeah, I know, after adding additional `afterException` and maybe `beforeException` fields to keep this information

Reply:
If we're only going to record one, the first one is probably more relevant - I'll let you make the call

Reply:
Right, let's keep it simple, I will set `exception` only if wasn't already set